samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java

```diff
@@ -274,6 +283,20 @@ void doOnProcessorChange() {
     LOG.info("Generating new JobModel with processors: {}.", currentProcessorIds);
     JobModel newJobModel = generateNewJobModel(processorNodes);
 
+    /*
+     * Leader skips the rebalance even if there are changes in the quorum as long as the work assignment remains the same
+     * across all the processors. The optimization is useful in the following scenarios
+     *   1. The processor in the quorum restarts within the debounce window. Originally, this would trigger rebalance
+     *      across the processors stopping and starting their work assignment which is detrimental to availability of
+     *      the system. e.g. common scenario during rolling upgrades
+     *   2. Processors in the quorum which don't have work assignment and their failures/restarts don't impact the
+     *      quorum.
+     */
+    if (newJobModel.equals(activeJobModel)) {
```

Comment:
This check would also include both the config and work-assignments for verifying equality between two different JobModels. 

Standalone allows users to launch two processors with different configs in a quorum. Would be better to narrow down this check to work-assignment alone to achieve the desired effect.

Reply:
Can we use the compareContainerModelForProcessor added in this PR for comaprison here: https://github.com/apache/samza/pull/1478/files

Reply:
Good point. I suppose the reason it works during my testing is `generateNewJobModel` doesn't store the configuration with the job model generated and hence the check happens to be the same as comparing the container model for all the processors. 

I will make it explicit to avoid this implicit assumption.